src/cluster_crypto.rs

```diff
@@ -490,13 +495,59 @@ impl ClusterCryptoObjects {
         self.establish_relationships().context("establishing relationships")?;
         log::info!("Established relationships between crypto objects");
 
+        if crypto_customizations.ip_change_only {
+            ensure!(
+                !crypto_customizations.cn_san_replace_rules.0.is_empty(),
+                "ip-change-only requires at least one cn_san_replace rule"
+            );
+            self.prune_cert_key_pairs_to_changed_cn_san_trees(&crypto_customizations.cn_san_replace_rules)
+                .context("pruning cert-key-pairs to CN/SAN-changed trees")?;
+        }
+
         self.regenerate_crypto(rsa_pool, crypto_customizations)
             .context("regenerating crypto")?;
         log::info!("Regenerated all crypto objects");
 
         Ok(())
     }
 
+    fn prune_cert_key_pairs_to_changed_cn_san_trees(&mut self, cn_san_replace_rules: &CnSanReplaceRules) -> Result<()> {
+        // For each root cert-key-pair tree: if *any* cert in the tree would change if
+        // cn_san_replace_rules were applied (CN and/or SAN), keep the *entire* tree.
+        let mut keep: HashSet<usize> = HashSet::new();
+
+        let roots: Vec<Rc<RefCell<CertKeyPair>>> = self
+            .cert_key_pairs
+            .iter()
+            .filter(|pair| (**pair).borrow().signer.is_none())
+            .cloned()
+            .collect();
+
+        // TODO: Do both check and mark in one pass
+        for root in roots {
+            let mut visited = HashSet::new();
```

Comment:
To avoid the awkward interface where the caller must supply the mutable visited set we should wrap the recursive function with another function that handles this implementation detail